### DISCUSSION TITLE
Message Boundaries and Parsing

### DISCUSSION THREAD

**Discussion by naveedpash** [librasn/rasn#52] [Q&A] [answered]
Hi,

I'm not a network programmer or a professional programmer for that matter. I am trying to implement CMIP over TCP/IP (CMOT) [(RFC 1189)](https://datatracker.ietf.org/doc/html/rfc1189) as an intermediate step in the implementation of IEEE 11073, a standard for communication to, from and between medical devices.

The `rasn` crate has been a great help in easing the development process. I have a question: Once I have implemented any particular RFC using `rasn`, how do I parse ASN.1 messages sent over a TCP connection? I understand that TCP is a stream oriented connection and not a packet oriented connection. So I guess I should ingest the first couple of octets to see the tag of the message, followed by another couple of octets to see the length of the message followed by ingesting the exact length that I just found out?

**Comment by XAMPPRocky** [comment, marked as answer]
Thank you for your question! Yeah, there should probably be an example of how to do this in the documentation, but what you need to do is use `rasn::ber::de::Decoder` directly, which has the [`decoded_len`](https://docs.rs/rasn/0.4.3/rasn/ber/de/struct.Decoder.html#method.decoded_len) method, which returns how many octets have been parsed. So you can use that get any remaining octets in the input with `input[decoder.decoded_len()..]`, and combine that with the new packet.

**Comment by naveedpash** [comment]
Alright, thank you for that. I'll try it out and if it works then I'll use my experience to send you a pull request to the documentation containing an example.

I have a follow up question: the CMOT standard (RFC 1189) has a bunch of different types of messages that can be sent. After using `decoded_len` to parse any one message, how do I know the type of the message has been parsed? Apart from the fields present in the messages, there's nothing in them that will distinguish one message type from another. Will I have to write an elaborate pattern matching function that will "guess" the type of message that has been received?

**Reply by XAMPPRocky** [reply]
> how do I know the type of the message has been parsed? 

Looking at RFC 1085, you'd use the top-level `PDUs` type, which will give you a choice that will tell you what kind of message it was. Then you can `match`, `if let`, or `matches!` to check what variant it is.

```rust
#[derive(AsnType, Encode, Decode)]
#[rasn(choice)]
pub enum Pdus {
    ConnectRequest(ConnectRequestPdu),
    ConnectResponse(ConnectResponsePdu),
    ReleaseRequest(ReleaseRequestPdu),
    ReleaseResponse(ReleaseResponsePdu),
    Abort(AbortPdu),
    UserData(UserDataPdu),
    ClUserData(ClUserDataPdu)
}
```

**Comment by naveedpash** [comment]
Oh my Lord, I cannot believe I've been staring at that RFC for a month and I didn't read that! **facepalm**

I was so absorbed in the original X.216 Presentation Protocol that forgot to scroll down the entire RFC...

Thank you @XAMPPRocky  you have made my year **thumbs up**
